Updated Widgets notebooks

JuanCab opened the pull request:
This is a first attempt at merging the widgets notebooks from Matt and Nicole's original widgets walkthrough.  I did restructure things a bit.  Still a few things to do here, so this is a draft PR for now.

Commit:
Clean up answers to nbdev notebook.

Commit:
Fixed typo in call for answer to import in 02a_nbdev notebook.

Commit:
Revised widgets notebook (not quite done) and its solutions

Commit:
Updated gitignore.

JuanCab committed:
Trimmed down the "Getting Started" to include no review of ipywidgets other than noting they exist and providing a display of all the widgets available.  This is because most of this is reviewed later.

JuanCab committed:
I changed some wording vis a vis the need for debuggers, since I use debuggers within Jupyter all the time.

JuanCab committed:
Updated data to current version.

JuanCab committed:
Moved widgets layout notebook from OldCode2PurgeEventually.

JuanCab committed:
Various bits of cleanup to widgets (saving work in progress).

JuanCab committed:
Turned updated data into proper CSV.

JuanCab committed:
Updated widgets to include ipydatagrid for rendering of Pandas dataframes.

JuanCab committed:
Updated the widgets notebook to add datagrid and updated all solutions.

JuanCab committed:
Change welcome message to reflect current plan.

Commit:
Final tweaks to widgets including solutions removal.

Commit:
Some cleanup to the nbdev and widgets notebook before cleanup of layout notebook.

Commit:
Cleaned up plot labels a bit and added additional documentation of where the data came from.

Commit:
Finished up the widgets layout and dashboard tests notebook cleanup.

Commit:
Fixed all the answer references and added a screenshot of the dashboard.

JuanCab commented:
OK, I have reworked the widgets notebooks.  Honestly, the widget content is very similar to before with a few nips and tucks.  I did spend some time adding more explanation of what was going on and I did use `ipydatagrid` to display the pandas data frame, since it stays one size.

Commit:
Merge remote-tracking branch 'Jupyter4Science/main' into widgets-tweaks

Commit:
Fixed name of dashboard test notebook in Voila notes.

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -150,7 +204,17 @@
    "source": [
     "#| export\n",
-    "year_range = widgets.IntRangeSlider(description = 'Range of Years')\n",
+    "year_range = widgets.IntRangeSlider(description = 'Range of Years')"
```
```suggestion
    "year_range = widgets.IntRangeSlider(description = 'Range of Years', style={'description_width': 'initial'})"
```

JuanCab replied:
Do we want to use this style here?  We introduce styling in the 02c_layout notebook and, in fact, THIS is one of the examples used in that notebook I think.

nicole-brewer replied:
I could go either way. I added it because it isn't just the label that is cut off, it is the numbers for me as well so it looked pretty bad

mwcraig replied:
Seems like it can't hurt to do it here and mention it again later.

mwcraig replied:
If numberes are getting cut off I think we're better off introducing this here and dropping it from the styling notebook later.

JuanCab replied:
OK, I will make the changes to both this notebook and the layout notebook.

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -380,25 +460,37 @@
    "outputs": [],
    "source": [
-    "year_range.value = (1990, 1993)"
+    "# The second (optional) argument to the observe method is the name of the trait\n",
+    "# to watch for changes. The default is 'All' which watches all traits.\n",
+    "# This particular widget only has a value trait, so asking to observe the 'value'\n",
+    "# trait is redundant... however, it is good practice to be explicit.\n",
+    "year_range.observe(on_range_change, 'value')"
```
```suggestion
    "# WARNING: The second argument to the observe method is the name of the trait\n",
    "# to watch for changes. The default is 'All' which watches all traits, including tr
    "# that start with an underscore that are intended only for internal use. \n",
    "# Watching all traits can be problematic for the same reason as in the function abo
    "# If too many changes occur at once, there is a chance your function will be called
    "# on a change to one of these internal traits - such as '_property_lock' - and cause
    "# your callback to fail or even silently cause unexpected behavior."
```

JuanCab replied:
Good point to include this.

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -358,19 +439,18 @@
+    "    global selected_df\n",
+    "    selected_df = original_df[(original_df['Year'] >= change['new'][0])\n",
+    "                              & (original_df['Year'] <= change['new'][1])]\n",
+    "    # NOTE: You could also use:\n",
+    "    #  selected_df = original_df[(original_df['Year'] >= year_range.value[0])\n",
+    "    #                             & (original_df['Year'] <= year_range.value[1])]\n",
+    "    # but it is better to use the 'change' object passed to the function because it is\n",
+    "    # not dependent on the name of the widget."
```
```suggestion
"    # WARNING: You could have thought to use:\n",
"    #  selected_df = original_df[(original_df['Year'] >= year_range.value[0])\n",
"    #                             & (original_df['Year'] <= year_range.value[1])]\n",
"    # but for many manual interactions, such as a click and drag slider, a high\n",
"    # overturn of quickly changing widget values could silently cause\n",
"    # unexpected behavior in changing values, so that syntax should be avoided"
```

JuanCab replied:
Interesting point!

mwcraig commented:
One quick question before I dive into a detailed review later today. Should the data be changing? I would think that we want the input temperature data to stay the same...

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -661,7 +826,7 @@
    "source": [
     "#| export\n",
-    "window_size = widgets.IntSlider(description = 'Window Size', min=1, max=100)"
+    "window_size = widgets.IntSlider(description = 'Window Size', value=20, min=1, max=100)"
```
```suggestion
    "window_size = widgets.IntSlider(description = 'Window Size', value=20, min=1, max=100, style={'description_width': 'initial'})"
```

JuanCab replied:
Same as before, do we want to hold off the introduction of styling until the layout notebook?

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -682,7 +855,7 @@
    "source": [
     "#| export\n",
-    "poly_order = widgets.BoundedIntText(description = 'Poly Order', min=0, max=10)"
+    "poly_order = widgets.BoundedIntText(description = 'Poly Order', min=0, value=3, max=10)"
```
```suggestion
    "poly_order = widgets.BoundedIntText(description = 'Poly Order', min=0, value=3, max=10, style={'description_width': 'initial'})"
```

JuanCab replied:
As before, styling should maybe wait...

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -836,13 +1024,17 @@
+    "# %answer key/dashboard/widgets.py 89\n",
     "\n",
     "def on_window_size_change(change):\n",
     "    global original_data, selected, poly_order\n",
```
```suggestion
    "    global original_data, selected_df, poly_order\n",
```

JuanCab replied:
Good catch.

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -1027,7 +1231,7 @@
    "outputs": [],
    "source": [
-    "display_plot(None)"
+    "output_plot(None)"
```
This line doesn't make obvious changes to the notebook, because at least for me, the output widget isn't even in the view by the time I get to this cell. I suggest doing the implicit display of the output widget right before the call to `output_plot`. Right now it takes place further up, and I have to scroll to see the change.

You could add some text to a markdown cell like:
Let's make sure our new callback works when we call it explicitly. If we display `plot_view` again, we shouldn't see any output.

then a cell with `plot_view`

ANother markdown cell:
However, when we call `output_plot` function, we should see the plot displayed with the `plot_view` output widget above.

a cell with `output_plot(None)`

JuanCab replied:
OK, I have tried to clean this up a bit as you suggested.  Should have this fix committed shortly.

JuanCab commented:
> One quick question before I dive into a detailed review later today. Should the data be changing? I would think that we want the input temperature data to stay the same...

All I did was update the datafile to the most recent version of the NASA data, so it runs through 2023 instead of 2020.  Other than updating the datafile, there is no 'change to the input data' in the sense of changing data mid-processing.  

Please clarify if I am not understanding your concern.

nicole-brewer commented on 02b_widgets.ipynb:
```diff
@@ -5,106 +5,145 @@
    "id": "e5c6ecde-5850-4d35-903e-6ed258669fe6",
    "metadata": {},
-   "outputs": [],
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "UsageError: Line magic function `%answer` not found.\n"
+     ]
+    }
+   ],
```
Just a friendly reminder that doing a "Restart and Run all cells" before commit makes it easier to review the diff, which I'm sure you know already and I forget frequently as well :)

JuanCab replied:
Dang, sorry about that. :)

nicole-brewer reviewed (changes requested):
Looks great! Minor suggestions only. Tried out the nbdev and widgets notebooks in full, including the use of the answer key with no problems.

JuanCab committed:
Update 02b_widgets.ipynb

Added clearer warning about the dangers of watching 'All' traits for changes.

Co-authored-by: Nicole Brewer <nbrewer6@asu.edu>

JuanCab committed:
Fixed unfixed reference to old data frame name.

Co-authored-by: Nicole Brewer <nbrewer6@asu.edu>

JuanCab committed:
Clarify comments about calling widget values being less than desirable.

Added Nicole's clarification of the dangers of calling values of widgets directly from a handler function.

Co-authored-by: Nicole Brewer <nbrewer6@asu.edu>

Commit:
Cleanup of some of the widgets notebook in response to Nicole's comments about showing plot_view Output widget closer to where it is changed interactively.

Commit:
Added a note after first slider displayed to note text may be cut off.

Commit:
Adjusted widgets and layout notebooks to handle widget styling a bit earlier.

JuanCab commented:
OK, I pushed some updates to address all of Nicole's comments on the widgets notebook including:
- Using style in widgets notebook so that the widgets are displayed with full descriptive text.
- Moving `plot_view` widget closer to where we manipulate it so we can see the output when updated.

mwcraig commented:
> All I did was update the datafile to the most recent version of the NASA data, so it runs through 2023 instead of 2020. 

Ah, I see that now. Not surprised that some of the earlier data was updated too -- I was originally thrown off by changes in data as far back as 1880. 

Definitely good to update with the latest data.

mwcraig commented on 01a_getting_started.ipynb:
```diff
@@ -125,365 +117,42 @@
+    "groups = organized_widgets(organize_by='ui')\n",
+    "help_url_base='reference/complete-ipywidgets-widget-list.ipynb'\n",
+    "list_overview_widget(groups, columns=2, min_width_single_widget=200, help_url_base=help_url_base)"
    ]
```
Maybe add a brief exercise here asking them to pick a widget (any slider, dropdown or text entry box) and ask them to make and display it. Should just be a copy/paste from the linked documentation.

mwcraig commented on 01b_workflow.ipynb:
```diff
@@ -29,7 +29,7 @@
     "We will build the web app using a Jupyter notebook where our workflow will:\n",
-    "- **load a data file** with global temperature data and\n",
+    "- **load a data file** with global land-ocean temperature data and\n",
     "- **smooth the data with several algorithms**.\n",
```
Do we use several algorithms? I'm happy to go there, especially as an exercise at the end, but I think right now there is only one smoothing algorithm.

mwcraig commented on 01b_workflow.ipynb:
```diff
@@ -140,8 +141,8 @@
+    "plt.title('Global Annual Mean Surface Air Temperature Change')\n",
     "plt.plot(df['Year'], df['Temperature'], label='Raw Data')\n",
     "plt.plot(df['Year'], df[SG_col], label='Savitzky-Golay Filtered')\n",
     "plt.plot(df['Year'], df[moving_avg_col], label='Moving Average Filtered')\n",
```
How about adding a `plt.grid()` too? I really like grids....

JuanCab commented:
> Ah, I see that now. Not surprised that some of the earlier data was updated too -- I was originally thrown off by changes in data as far back as 1880.

Yeah, I noticed that too but assumed some re-reduction of old data had tweaked the values.

mwcraig commented on 02a_nbdev.ipynb:
```diff
@@ -88,7 +86,9 @@
+    "> **IMPORTANT:** Two additional hints about using nbdev to create a Python module:\n",
+    "> 1.  You can also give it a “dotted module names” as is done with Python packages. For example, a `#| default_exp` directive with an output file of `some.module` will create a file called `some/module.py` within the package directory.\n",
+    "> 2.  Each Jupyter notebook can only be exported to a SINGLE Python module. If you want to export cells to multiple modules, you will need to create multiple notebooks."
    ]
```
How about adding a number 3 too:

3. Only changes to the notebook that have been saved to disk will be written to the Python file.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -169,23 +227,25 @@
-    "That looks right. Let's use the same syntax to set the min and max of the slider widget."
+    "Widgets, like other Python objects, have attributes. Recall that to set the value of an attribute of an object, we use 'dot notation' (e.g. `box.width `sets the `width` attribute of object `box`).\n",
+    "\n",
+    "Set the `min` and `max` attributes of the slider widget.  Then look back at the widget"
```
Can we add some kind of indicator that this is an exercise? I've been having a little trouble following when I'm supposed to do something vs just run the code already there.

mwcraig replied:
Maybe mention that the order matters: if you try to set the minimum first there will be an exception because the minimum you are trying to set is larger than the maximum.

I think it wopuld also be fine to not mention it and just discuss it when it comes up in the room.

mwcraig commented on key/dashboard/widgets.py:
```diff
@@ -1,108 +1,124 @@
-year_range.max =  max(original_data['Year'])  # set the 'max' attribute of the slider to the minimum year of the our data  
-year_range.min = min(original_data['Year'])  # and let's do the same for 'min'
+# %% ../02b_widgets.ipynb 20
+year_range.max =  max(original_df['Year']) # set the 'max' attribute of the slider to the minimum year of the our data
```
```suggestion
year_range.max =  max(original_df['Year']) # set the 'max' attribute of the slider to the maximum year of the our data
```

mwcraig replied:
I think this needs to be changed

mwcraig replied:
The change here is "minimum" → "maximum" in the comment.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -150,7 +195,20 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Display the widget\n",
```
Maybe suggest in the text that they make a new view for this cell so it is easier to see what happens as they make the changes in the next few cells.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -245,11 +307,11 @@
     "\n",
-    "## Use the slider value to update the `selected` data\n",
-    "Well, we probably want to keep two copies, because if we change the original to have a shorter range, and then we want to expand that range, we're going to have lost data. So let's maintain a dataframe called `selected` that reflects the range specified by the Slider. Let's look at how we did that in workflow.ipynb. \n",
+    "## Use the slider value to update the `selected_df` data\n",
```
```suggestion
    "## Use the slider value to update the selected data\n",
```

I'd leave the variable name out here since the variable doesn't exist yet, and it reads a little more naturally to me if the title explains in plain words what the goal is.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -245,11 +307,11 @@
-    "## Use the slider value to update the `selected` data\n",
-    "Well, we probably want to keep two copies, because if we change the original to have a shorter range, and then we want to expand that range, we're going to have lost data. So let's maintain a dataframe called `selected` that reflects the range specified by the Slider. Let's look at how we did that in workflow.ipynb. \n",
+    "## Use the slider value to update the `selected_df` data\n",
+    "Well, we probably want to keep two copies, because if we change the original to have a shorter range, and then we want to expand that range, we're going to have lost data. So let's maintain a dataframe called `selected_df` that reflects the range specified by the Slider. Let's look at how we did that in workflow.ipynb. \n",
```
```suggestion
    "Well, we probably want to keep two copies, like we did in the workflow notebook, because if we change the original to have a shorter range, and then we want to expand that range, we're going to have lost data. So let's maintain a dataframe called `selected_df` that reflects the range specified by the Slider. Let's look at how we did that in workflow.ipynb. \n",
```

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -539,118 +652,178 @@
-    "#| export\n",
+    "# Create a handler function that will display the selected dataframe\n",
+    "def output_selected_dataframe(change):\n",
+    "    selected_dataframe_view.clear_output(wait=True)\n",
```
We should explain somewhere what the `wait` argument does (delays erasing the old output until the new output is ready to display) -- it avoids flickering as the output changes.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -460,9 +557,9 @@
    "source": [
     "#| export\n",
-    "# %answer key/dashboard/widgets.py  40\n",
+    "# %answer key/dashboard/widgets.py  46\n",
```
When I run this cell it loads all of widgets.py.....

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -758,7 +955,9 @@
-    "Sweet! If you still have an Output View of the `selected` dataframe open, and you increase the value of `window_size` you will see that there is now a Savitzky-Golay column. Now I want you to create a Slider widget that is linked to the `poly_order`. "
+    "Sweet! If you still have an Datagrid View of the `selected_df` dataframe open, and you increase the value of `window_size` you will see that there is now a \"Smoothed Data\"\" column. \n",
+    "\n",
+    "Now create a Slider widget that is linked to the `poly_order`. "
```
We've already created the `poly_order` widget above, and made it a `BoundedIntBox`. Here I think it is only observer that needs to be changed.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -539,118 +652,178 @@
-    "    selected_data_output.clear_output(wait=True)\n",
-    "    with selected_data_output: \n",
-    "        display(selected)"
+    "Let's connect the `on_range_change` function to the `year_range` widget with the `observe` method."
```
```suggestion
    "Let's connect the `output_selected_dataframe` function to the `year_range` widget with the `observe` method."
```

mwcraig replied:
This still needs to be fixed, I think...

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -539,118 +652,178 @@
    "metadata": {},
    "source": [
-    "Way cool. Now our dataframe reflects the range of values."
+    "Let's create yet another copy of the `year_range` slider and see how changing it affects the datagrid widget we created above."
```
Why not just make a new view for the slider?

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -0,0 +1,752 @@
+   "id": "ae93f0c4",
+   "metadata": {},
+   "source": [
+    "While we only used one very utilitarian style attribute, but it is worth noting that you can use the style attribute to change your widgets in a variety of aesthetic ways: changing their colors, fonts, and more. A full list of which attributes are accessible for each widget is [available in the Jupyter widgets documentation](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Styli ...
```
```suggestion
    "While we only used one very utilitarian style attribute, it is worth noting that you can use the style attribute to change your widgets in a variety of aesthetic ways: changing their colors, fonts, and more. A full list of which attributes are accessible for each widget is [available in the Jupyter widgets documentation](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Styling.html#current-supported-attributes).\n"
```

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -0,0 +1,752 @@
+   "source": [
+    "#| export\n",
+    "selected_data_accordion = widgets.Accordion(titles=('Selected Data',))\n",
+    "selected_data_accordion"
```
Have them make a new view for this output....

mwcraig replied:
This still might be useful

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -0,0 +1,752 @@
+   "id": "e9d46022-ef0a-4d0b-bb46-033cd9f8330b",
+   "metadata": {},
+   "source": [
+    "### Layout widget objects\n",
```
I don't think we need to point this out, but the `Layout` is itself a widget, so you can observe changes in its properties if you want.

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -0,0 +1,752 @@
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "intro_text = widgets.HTML(value = INTRO_TEXT, layout = widgets.Layout(max_width = '500px'))\n",
```
```suggestion
    "html_layout = widgets.Layout(max_width = '500px')\n",
    "intro_text = widgets.HTML(value=INTRO_TEXT, layout=html_layout)\n",
```

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -0,0 +1,752 @@
+   "source": [
+    "#| export\n",
+    "intro_text = widgets.HTML(value = INTRO_TEXT, layout = widgets.Layout(max_width = '500px'))\n",
+    "data_source_text = widgets.HTML(value = SOURCES_TEXT, layout = widgets.Layout(max_width = '500px'))"
```
```suggestion
    "data_source_text = widgets.HTML(value = SOURCES_TEXT, layout = html_layout)"
```

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -0,0 +1,752 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "poly_order"
```
Why are `window_size` and `poly_order` displayed again here? Seems like making the `HBox` is the next step....

mwcraig reviewed (changes requested):
Overall this is looking good! I made a number of line comments/suggestions and have a few more comments below on specific things I couldn't add a line comment on.

I was a little confused throughout on whether the cells with answers already had the answer loaded, but I think in those cases code was there with comments that were supposed to be replaced by the user.

+ Does the `%export` magic get used anywhere? Could we leave it out? 
+ In 02a_nbdev a few suggestions/comments
    - I found I needed to export the imports at the top, the first time they are loaded, rather than the cell of imports that comes after the definition of the `answer` magic. If I didn't do that then the next notebook raises an error on importing `dashboard` because `register_cell_magic` is being used before it is imported. 
    - Maybe move the paragraph that starts "Isn't nbdev great!" to a later notebook, after we have finished building the dashboard? It seems premature at this stage. `02d_dashboard` might be a good place to put it.
    - In the first place where you use the `%answer` magic point out that you need to execute that cell *twice*: once to load the solution and again to run the solution. One visual cue that it needs to be rerun is that the cell label turns yellow.
+ In `02b_widgets` should the minimum window size be 2 instead of 1?   I guess this would apply earlier too.

JuanCab commented:
> * In `02b_widgets` should the minimum window size be 2 instead of 1?   I guess this would apply earlier too.

I think the smoothing algorithm does fit with order 0, but it doesn't do anything.  I'll see about making this change throughout.

Commit:
Updated .gitignore to not see voici build output.

Commit:
Updated all the notebooks and solutions to reflect @mattwcraig suggestions.

JuanCab commented:
Matt's issues that I have addressed (since the PR comments interface isn't letting me address these individually):

>In 02a_nbdev a few suggestions/comments
> - I found I needed to export the imports at the top, the first time they are loaded, rather than the cell of imports that comes after the definition of the answer magic. If I didn't do that then the next notebook raises an error on importing dashboard because register_cell_magic is being used before it is imported.

Agreed, and fixed.

> Maybe move the paragraph that starts "Isn't nbdev great!" to a later notebook, after we have finished building the dashboard? It seems premature at this stage. 02d_dashboard might be a good place to put it.

Done.

> In the first place where you use the %answer magic point out that you need to execute that cell twice: once to load the solution and again to run the solution. One visual cue that it needs to be rerun is that the cell label turns yellow.

Done.

> In `02b_widgets` should the minimum window size be 2 instead of 1? I guess this would apply earlier too.

I changed both `window_size` and `poly_order` to reflect an updated idea that the polynomial order should be at least 1 which mandates a window size of 2 or above.

> Does the %export magic get used anywhere? Could we leave it out?

It is not used anywhere I can see, it looks like it may have been meant by @nicole-brewer as a way of exporting notebooks in nbdev with solutions, but not sure...

> Maybe add a brief exercise here asking them to pick a widget (any slider, dropdown or text entry box) and ask them to make and display it. Should just be a copy/paste from the linked documentation.

Done.

> Do we use several algorithms? I'm happy to go there, especially as an exercise at the end, but I think right now there is only one smoothing algorithm.

I reworded this to state "smooth the data with an algorithm with different smoothing parameters"

> How about adding a number 3 too:
> 3. Only changes to the notebook that have been saved to disk will be written to the Python file..

Done.

> Can we add some kind of indicator that this is an exercise? I've been having a little trouble following when I'm supposed to do something vs just run the code already there.

I have added an **EXERCISE** (all caps) to indicate exercises (I did this to all these notebooks).  In the process I removed the answers that I forgotten to remove from `02c_layout.ipynb` 🤦🏼

> Maybe mention that the order matters: if you try to set the minimum first there will be an exception because the minimum you are trying to set is larger than the maximum.
> I think it wopuld also be fine to not mention it and just discuss it when it comes up in the room.

Let's wait and see if this happens to address it.

> I don't think we need to point this out, but the Layout is itself a widget, so you can observe changes in its properties if you want.

I added a "PRO TIP" about this...

> Why are window_size and poly_order displayed again here? Seems like making the HBox is the next step....

Removed them.

mwcraig commented on 02a_nbdev.ipynb:
```diff
@@ -326,14 +332,14 @@
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 9,
```
Output should be cleared here

JuanCab replied:
Sigh, I keep forgetting that. :(

mwcraig commented on 02a_nbdev.ipynb:
```diff
@@ -359,7 +365,7 @@
     "\n",
-    "Another way to verify that this worked is to import the function we just wrote to the Python file. Can you guess what that import statement would look like?"
+    "**EXERCISE**: Another way to verify that this worked is to import the function we just wrote to the Python file. Can you guess what that import statement would look like?"
    ]
```
The code cell below needs to have output removed

JuanCab replied:
Wait a sec, I don't see any output in my commit of that cell...  Mine shows:
```
# %answer key/02a/01.py

# import ...
```

mwcraig commented on 02a_nbdev.ipynb:
```diff
@@ -106,7 +109,9 @@
-    "  - The great benefit of using nbdev for development can also be the most challenging part of using nbdev - making sure we have not forgotten to export all the right cells to our resulting Python file. If you ever get a confusing or unexpected error on importing a module created with nbdev, it's very likely you forgot to export one of your code cells."
+    "  - The great benefit of using nbdev for development can also be the most challenging part of using nbdev - making sure we have not forgotten to export all the right cells to our resulting Python file. If you ever get a confusing or unexpected error on importing a module created with nbdev, it's very likely you forgot to export one of your code cells.\n",
+    "\n",
+    "**EXERCISE**: Add the missing export directive to export this cell to the dashboard module to complete the following cell..."
```
Either don't make this an exercise -- which would be fine so that they see a first example of export -- or remove the `#| export` from the cell below.

JuanCab replied:
Fixed.

mwcraig commented on 02a_nbdev.ipynb:
```diff
@@ -381,12 +387,14 @@
-    "Great job! Now we know the basic idea of what nbdev does. Go ahead and use the `%load` line magic inspect the exported file below."
+    "Great job! Now we know the basic idea of what nbdev does. \n",
+    "\n",
+    "**EXERCISE**: Go ahead and use the `%load` line magic inspect the exported `dashboard.__init__` file below."
```
```suggestion
    "**EXERCISE**: Go ahead and use the `%load` line magic to inspect the exported `dashboard.__init__` file below."
```

JuanCab replied:
Fixed

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -5,89 +5,88 @@
+    "### Steps You Will Take In This Notebook\n",
+    "In this notebook you will:\n",
+    "1. Briefly review ipywidgets.\n",
+    "2. Load data from file.\n",
```
```suggestion
    "2. Load data from a file.\n",
```

JuanCab replied:
Fixed.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -98,13 +97,44 @@
+    "+ A communication framework between the front end (your browser) and back end (Python or other kernel).\n",
+    "+ A set of fundamental user interface elements like buttons and checkboxes.\n",
+    "\n",
+    "Execute the next cell create a browser of the available elements. To see some example code for implementing any of the elements click on its title. It will be easier to view both this overview and [the code examples](reference/complete-ipywidgets-widget-list.ipynb) if you have them open in separate tabs.\n",
```
"Execute the next cell create" → "Execute the next cell **to** create"

JuanCab replied:
Done.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -358,19 +434,31 @@
-   "outputs": [],
    "source": [
-    "year_range.observe(on_range_change, 'value')"
+    "And now let's test it by directing the `observe` method to call this new `on_range_change` every time the value of the slider changes. While we do this, let's address a poor Jupyter widget programming practice we showed earlier.  There is an optional second argument to the `observe` method which is the name of the trait to watch for changes.  We didn't include it... this is not a good choice ...
```
I would suggest adding a line break after the first sentence to make it clearer that you want them to try out the slider.

JuanCab replied:
Done.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -460,9 +558,9 @@
    "source": [
     "#| export\n",
-    "# %answer key/dashboard/widgets.py  40\n",
+    "# %answer key/dashboard/widgets.py  46\n",
```
Need to change `widgets.py  46` to `widgets.py 46`. The first one has *two* spaces, which confuses the answer magic, so it loads the whole file instead of just the single line from cell 46.

JuanCab replied:
I wonder if I could fix the magic... but for now, space removed.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -539,118 +653,178 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "selected_data_grid"
```
I would add a markdown cell above this suggesting that they make a new view for this output too.

JuanCab replied:
Fixed both of these I think.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -716,20 +900,24 @@
-    "     selected = original_data[(original_data['Year'] >= year_range.value[0]) & (original_data['Year'] <= year_range.value[1])]"
+    "    global original_df, selected_df # Remember to access the global variables\n",
+    "    # catch the change in the window_size widget value and update the original_df\n",
+    "    # original_df['Smoothed Data'] = savgol_filter(original_df['Temperature'],\n",
```
I would suggest shortening the comment hint to something like

```suggestion
    "    # original_df['Smoothed Data'] = savgol_filter(...,\n",
```

and then delete the next few lines with the argument to `savgol_filter`. I found it hard to spot the change I was supposed to make.

JuanCab replied:
The point is the comment
```
    # original_df['Smoothed Data'] = savgol_filter(original_df['Temperature'],
    #                                              window_size,
    #                                              poly_order).round(decimals=3)
``` 
is correct except for the window_size argument, which they need to fix.  I tried to change the comment in front of this to make it clearer.  I also removed the `original_df` call from the comments and just inserted it.

mwcraig replied:
It isn't correct, though -- `poly_order` should be `poly_order.value`, which it is in the solution, but not in the stareting text.

JuanCab replied:
This is intentional, I wanted the person taking the tutorial to realize `window_size` is handled with change but `poly_order` isn't in this scenario.  The comments tried to make this clear.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -716,20 +900,24 @@
    "metadata": {},
    "source": [
-    "Then do you recall what function we used to display the dataframe? It was `display_selected_data`. Let's make sure we add both to the functions observed by window_size."
+    "Then do you recall what function we used to display the dataframe? It was `display_selected_dataframe`. Let's make sure we add both to the functions observed by `window_size`."
```
```suggestion
    "Then do you recall what function we used to display the dataframe? It was `update_selected_datagrid`. Let's make sure we add both to the functions observed by `window_size`."
```

Not sure if this is the right name, but `display_selected_dataframe` is definitely not the right one. Also wasn't clear what you meant by "both" until I read the next code cell

JuanCab replied:
I cleaned up the language to make all this clearer (I hope).

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -758,7 +956,9 @@
    "metadata": {},
    "source": [
-    "Sweet! If you still have an Output View of the `selected` dataframe open, and you increase the value of `window_size` you will see that there is now a Savitzky-Golay column. Now I want you to create a Slider widget that is linked to the `poly_order`. "
+    "Sweet! If you still have an Datagrid View of the `selected_df` dataframe open, and you increase the value of `window_size` you will see that there is now a \"Smoothed Data\"\" column. \n",
```
```suggestion
    "Sweet! If you increase the value of `window_size` you will see that there is now a \"Smoothed Data\"\" column. \n",
```

A view of the data grid was just created....

JuanCab replied:
OK.  Accepted.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -769,12 +969,13 @@
-    "    selected = original_data[(original_data['Year'] >= year_range.value[0]) & (original_data['Year'] <= year_range.value[1])]"
+    "    global original_df, selected_df\n",
+    "    # catch the change in the poly_order widget value and update the original_df\n",
+    "    # original_df['Smoothed Data'] = savgol_filter(original_df['Temperature'], window_size, poly_order).round(decimal=3)\n",
```
Also suggest shortening here to

```python
# original_df['Smoothed Data'] = savgol_filter(....
```

JuanCab replied:
Again, fixed by removing it as a comment and directing the user to edit the line.

mwcraig replied:
Note that another change needed here, which is in the solution but not the starting comment: `window_size` → `window_size.value`

JuanCab replied:
> Note that another change needed here, which is in the solution but not the starting comment: `window_size` → `window_size.value`

This was intentional, they have to change BOTH the poly_order and window_size... the approach is different depending on which we are monitoring for a 'change'.

mwcraig commented on key/dashboard/widgets.py:
```diff
@@ -1,108 +1,124 @@
+        plt.ylabel('Temperature Anomalies over Land w.r.t. 1951-80 (˚C)')
+        plt.title('Global Annual Mean Surface Air Temperature Change')
+        plt.plot(selected_df['Year'], selected_df['Temperature'], label='Raw Data')
+        plt.plot(selected_df['Year'], selected_df['Smoothed Data'], label='Smoothed Data')
```
Add a `plt.legend()` here

JuanCab replied:
Done.

mwcraig commented on key/dashboard/widgets.py:
```diff
@@ -1,108 +1,124 @@
-# %% ../../04_widgets.ipynb 76
+# %% ../02b_widgets.ipynb 89
 def on_window_size_change(change):
     global original_data, selected, poly_order
```
👀 this one is important -- without it the dashboard doesn't work, because `selected_df` is treated as a local variable. 

```suggestion
    global original_data, selected_df, poly_order
```

JuanCab replied:
Oh, when I fixed it in the notebook, I didn't catch it in the answer key. :(. Fixed.

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -0,0 +1,738 @@
+   "outputs": [],
+   "source": [
+    "#| export\n",
+    "# %answer key/dashboard/main.py 5\n",
```
I would drop this as an exercise...but if you want to keep then I would remove the answer from the cell 😀

JuanCab replied:
Fixed (by removing the answer).

mwcraig reviewed (changes requested):
This is looking really good! I have a number of small suggestions, including one critical error in `key/dashboard/widgets.py` that prevents the `window_size` and `poly_order` from changing the plot.

mwcraig commented:
>> Does the %export magic get used anywhere? Could we leave it out?

> It is not used anywhere I can see, it looks like it may have been meant by @nicole-brewer as a way of exporting notebooks in nbdev with solutions, but not sure...

There is a `-e` option to the `answer` magic that calls `export` so we should leave it in rather than trying to edit the code for the magics.

Commit:
Essentially addressed almost all suggestions by @mwcraig on Jun 25.

Commit:
Updated solutions to 02b_widgets.ipynb and tweaked some instructions.

Commit:
CLeaned up 02c_layout.ipynb and updated the solutions.

Commit:
Fixed solutions issues with 02c_layout

JuanCab commented:
OK, I have addressed all of @mwcraig recent feedback EXCEPT where not marked resolved.  It was a few small items.

Commit:
Another @mwcraig wording suggestion accepted.

Commit:
Fixed typo in comments.

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -81,10 +89,22 @@
   {
    "cell_type": "code",
-   "execution_count": null,
+   "execution_count": 6,
```
Need to clear output...

mwcraig commented on 02c_layout.ipynb:
```diff
@@ -81,10 +89,22 @@
+      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
+      "Cell \u001b[0;32mIn[6], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Display dataframe\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[43mselected_data_grid\u001b[49m\n",
+      "\u001b[0;31mNameError\u001b[0m: name 'selected_data_grid' is not defined"
```
This is a real error, will note the source of the error in 02b

JuanCab replied:
Where is this error?  Are you sure it is not because you don't have the assumed correct module files in the `dashboard` directory?  Copy the `key/dashboard` to `dashboard` and see if the result changes.  I am not seeing this error.

mwcraig replied:
Not sure what happened, but when I made sure to start with no `dashboard` directory when I started, it worked. Some of what went wrong, I think, is that some of the early exercises required adding nbdev directives that I might have missed.

mwcraig commented on key/02c/01.py:
```diff
@@ -1,3 +0,0 @@
-import ipywidgets as widgets
```
Was this file supposed to be deleted?

JuanCab replied:
Yeah, I created it temporarily when I thought I needed to use it, but the problem is if you host an answer in a non-module file, you can't successfully export the cell to nbdev.  So I removed this.

mwcraig commented on 02b_widgets.ipynb:
```diff
@@ -539,118 +656,204 @@
    "metadata": {},
    "outputs": [],
    "source": [
     "#| export\n",
```
There is an important error here -- `nbdev` cannot handle an import and other code in the same cell, so this needs to be broken into two `#| export` cells, on for the import and one for creating `selector_data_grid`

mwcraig replied:
Never mind, starting with a clean checkout worked.

mwcraig reviewed (changes requested):
A couple more changes, including an important one to fix a break of the dashboard, I think.

mwcraig reviewed (approved):
There may still a couple of minor things here or there, but I think this is ready to merge. Any remaining polishing will be easier in a new pull request.

JuanCab commented:
> There may still a couple of minor things here or there, but I think this is ready to merge. Any remaining polishing will be easier in a new pull request.

Thanks.  Merging.
